From: Justus Winter <justus@g10code.com>
Date: Tue, 13 Jun 2017 15:35:01 +0200
Subject: gpg: Check and fix keys on import.

* doc/gpg.texi: Document the new import option.
* g10/gpg.c (main): Make the new option default to yes.
* g10/import.c (parse_import_options): Parse the new option.
(import_one): Act on the new option.
* g10/options.h (IMPORT_REPAIR_KEYS): New macro.

GnuPG-bug-id: 2236
Originally-by: Justus Winter <justus@g10code.com>
(cherry picked from commit 9b12b45aa5e67d4d422bf75a3879df1d52dbe67f)

the original change depended on quite a bit of refactoring; my
cherry-pick ignores the refactoring and just applies to the code where
it originally sat.

This is aimed at resolving https://bugs.debian.org/906545

Signed-off-by: Daniel Kahn Gillmor <dkg@fifthhorseman.net>
---
 doc/gpg.texi  | 4 ++++
 g10/gpg.c     | 5 +++--
 g10/import.c  | 6 ++++++
 g10/keyedit.c | 2 +-
 g10/main.h    | 2 ++
 g10/options.h | 1 +
 6 files changed, 17 insertions(+), 3 deletions(-)

diff --git a/doc/gpg.texi b/doc/gpg.texi
index 9f95c9a6f..d08f6ac4d 100644
--- a/doc/gpg.texi
+++ b/doc/gpg.texi
@@ -2291,6 +2291,10 @@ opposite meaning. The options are:
   on the keyring. This option is the same as running the @option{--edit-key}
   command "clean" after import. Defaults to no.
 
+  @item repair-keys.  After import, fix various problems with the
+  keys.  For example, this reorders signatures, and strips duplicate
+  signatures.  Defaults to yes.
+
   @item import-minimal
   Import the smallest key possible. This removes all signatures except
   the most recent self-signature on each user ID. This option is the
diff --git a/g10/gpg.c b/g10/gpg.c
index 09e50db46..88c92dd59 100644
--- a/g10/gpg.c
+++ b/g10/gpg.c
@@ -2366,9 +2366,10 @@ main (int argc, char **argv)
     opt.max_cert_depth = 5;
     opt.escape_from = 1;
     opt.flags.require_cross_cert = 1;
-    opt.import_options = 0;
+    opt.import_options = IMPORT_REPAIR_KEYS;
     opt.export_options = EXPORT_ATTRIBUTES;
-    opt.keyserver_options.import_options = IMPORT_REPAIR_PKS_SUBKEY_BUG;
+    opt.keyserver_options.import_options = (IMPORT_REPAIR_KEYS
+					    | IMPORT_REPAIR_PKS_SUBKEY_BUG);
     opt.keyserver_options.export_options = EXPORT_ATTRIBUTES;
     opt.keyserver_options.options = KEYSERVER_HONOR_PKA_RECORD;
     opt.verify_options = (LIST_SHOW_UID_VALIDITY
diff --git a/g10/import.c b/g10/import.c
index a83551c41..6d62168b8 100644
--- a/g10/import.c
+++ b/g10/import.c
@@ -179,6 +179,9 @@ parse_import_options(char *str,unsigned int *options,int noisy)
        N_("assume the GnuPG key backup format")},
       {"import-restore", IMPORT_RESTORE, NULL, NULL},
 
+      {"repair-keys", IMPORT_REPAIR_KEYS, NULL,
+       N_("repair keys on import")},
+
       /* Aliases for backward compatibility */
       {"allow-local-sigs",IMPORT_LOCAL_SIGS,NULL,NULL},
       {"repair-hkp-subkey-bug",IMPORT_REPAIR_PKS_SUBKEY_BUG,NULL,NULL},
@@ -1471,6 +1474,9 @@ import_one (ctrl_t ctrl,
     log_info (_("key %s: PKS subkey corruption repaired\n"),
               keystr_from_pk(pk));
 
+  if ((options & IMPORT_REPAIR_KEYS))
+    check_all_keysigs (keyblock, 0, 0);
+
   if (chk_self_sigs (keyblock, keyid, &non_self))
     return 0;  /* Invalid keyblock - error already printed.  */
 
diff --git a/g10/keyedit.c b/g10/keyedit.c
index 69946ea13..3d6c5d4c4 100644
--- a/g10/keyedit.c
+++ b/g10/keyedit.c
@@ -387,7 +387,7 @@ sig_comparison (const void *av, const void *bv)
    however).
 
    Returns 1 if the keyblock was modified, 0 otherwise.  */
-static int
+int
 check_all_keysigs (KBNODE kb, int only_selected, int only_selfsigs)
 {
   gpg_error_t err;
diff --git a/g10/main.h b/g10/main.h
index edee02708..8f89612ed 100644
--- a/g10/main.h
+++ b/g10/main.h
@@ -300,6 +300,8 @@ void keyedit_quick_sign (ctrl_t ctrl, const char *fpr,
 void keyedit_quick_set_expire (ctrl_t ctrl,
                                const char *fpr, const char *expirestr);
 void show_basic_key_info (KBNODE keyblock);
+int check_all_keysigs (KBNODE kb, int only_selected, int only_selfsigs);
+
 
 /*-- keygen.c --*/
 const char *get_default_pubkey_algo (void);
diff --git a/g10/options.h b/g10/options.h
index fda174f80..059782416 100644
--- a/g10/options.h
+++ b/g10/options.h
@@ -352,6 +352,7 @@ EXTERN_UNLESS_MAIN_MODULE int memory_stat_debug_mode;
 #define IMPORT_KEEP_OWNERTTRUST          (1<<8)
 #define IMPORT_EXPORT                    (1<<9)
 #define IMPORT_RESTORE                   (1<<10)
+#define IMPORT_REPAIR_KEYS               (1<<11)
 
 #define EXPORT_LOCAL_SIGS                (1<<0)
 #define EXPORT_ATTRIBUTES                (1<<1)
